sklearn/utils/tests/test_stats.py

```diff
@@ -1,43 +1,70 @@
 import numpy as np
+import pytest
 from numpy.testing import assert_allclose
 from pytest import approx
 
+from sklearn._config import config_context
+from sklearn.utils._array_api import (
+    yield_namespace_device_dtype_combinations,
+)
+from sklearn.utils.estimator_checks import _array_api_for_tests
 from sklearn.utils.stats import _weighted_percentile
 
 
-def test_weighted_percentile():
-    y = np.empty(102, dtype=np.float64)
-    y[:50] = 0
-    y[-51:] = 2
-    y[-1] = 100000
-    y[50] = 1
-    sw = np.ones(102, dtype=np.float64)
+@pytest.mark.parametrize(
+    "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
+)
+def test_weighted_percentile(array_namespace, device, dtype_name):
+    xp = _array_api_for_tests(array_namespace, device)
+
+    y_np = np.empty(102, dtype=dtype_name)
+    y_np[:50] = 0
+    y_np[-51:] = 2
+    y_np[-1] = 100000
+    y_np[50] = 1
+    y = xp.asarray(y_np, device=device)
+    sw_np = np.ones(102, dtype=dtype_name)
+    sw = xp.asarray(sw_np, device=device)
     sw[-1] = 0.0
-    score = _weighted_percentile(y, sw, 50)
-    assert approx(score) == 1
-
-
-def test_weighted_percentile_equal():
-    y = np.empty(102, dtype=np.float64)
-    y.fill(0.0)
-    sw = np.ones(102, dtype=np.float64)
+    # score = _weighted_percentile(y, sw, 50)
+    # assert approx(score) == 1
```

Comment:
Regarding `array_api_dispatch` and a problem I noticed. When `array_api_dispatch` is not turned on, `xp` is always a NumPy wrapper. This is a problem when the actual namespace is Torch, because the NumPy wrapper tricks the code into thinking that the namespace is NumPy and try to use `np` functions, which will result in the NumPy not available error. I am not sure how to get around this, so any advice would be appreciated.

Reply:
> This is a problem when the actual namespace is Torch, because the NumPy wrapper tricks the code into thinking that the namespace is NumPy and try to use np functions, which will result in the NumPy not available error.

I don't understand what you mean by "NumPy not available error". Would you mind quoting the specific traceback and exception message you observed?

When array API dispatch, scikit-learn should ideally behave at it used to prior we started introducing array API support:

- pytorch tensors allocated on a cpu device would be transparently converted to numpy arrays since they implement the `__array__` protocol (much older than and unrelated to array API):

```python
>>> import numpy as np
>>> import torch
>>> from sklearn.utils.validation import check_array
>>> a = torch.tensor([0., 1., 2.])
>>> a
tensor([0., 1., 2.])
>>> a.device
device(type='cpu')
>>> np.asarray(a)
array([0., 1., 2.], dtype=float32)
>>> check_array(a, ensure_2d=False)
array([0., 1., 2.], dtype=float32)
```

- tensor allocated on a non-CPU device (e.g. "cuda" or "mps") would raise an error when attempting to convert to a numpy array:

```python
>>> b = a.to("mps")
>>> b
tensor([0., 1., 2.], device='mps:0')
>>> check_array(b, ensure_2d=False)
Traceback (most recent call last):
  Cell In[24], line 1
    check_array(b, ensure_2d=False)
  File ~/code/scikit-learn/sklearn/utils/validation.py:1052 in check_array
    array = _asarray_with_order(array, order=order, dtype=dtype, xp=xp)
  File ~/code/scikit-learn/sklearn/utils/_array_api.py:883 in _asarray_with_order
    array = numpy.asarray(array, order=order, dtype=dtype)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/torch/_tensor.py:1087 in __array__
    return self.numpy()
TypeError: can't convert mps:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

Reply:
Here is the error message and traceback I got. Looking a bit closer at it, I think it may be because `xp, is_array_api_compliant, device = get_namespace_and_device(array)` returned `None` as device when the device should have been `cpu`. (I used it [here](https://github.com/scikit-learn/scikit-learn/pull/29431/files#diff-c7d6538a14900fc61a71b07411b38f654ed7e35f54e1150ca5e73ae31cab7c0eR40)) I wonder if I am using this function incorrectly... 

```
sklearn/utils/tests/test_stats.py::test_weighted_percentile[torch-cpu-float64] FAILED                                                                                                                    [  7%]

=================================================================================================== FAILURES ===================================================================================================
_________________________________________________________________________________ test_weighted_percentile[torch-cpu-float64] __________________________________________________________________________________

array_namespace = 'torch', device = 'cpu', dtype_name = 'float64'

    @pytest.mark.parametrize(
        "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
    )
    def test_weighted_percentile(array_namespace, device, dtype_name):
        xp = _array_api_for_tests(array_namespace, device)
    
        y_np = np.empty(102, dtype=dtype_name)
        y_np[:50] = 0
        y_np[-51:] = 2
        y_np[-1] = 100000
        y_np[50] = 1
        y = xp.asarray(y_np, device=device)
        sw_np = np.ones(102, dtype=dtype_name)
        sw = xp.asarray(sw_np, device=device)
        sw[-1] = 0.0
>       score = _weighted_percentile(y, sw, 50)

array_namespace = 'torch'
device     = 'cpu'
dtype_name = 'float64'
sw         = tensor([0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000,
        1.8750, 0.0000, 1.8750, 0.0000....0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750,
        0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 0.0000])
sw_np      = array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., ...1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.])
xp         = <module 'array_api_compat.torch' from '/Users/emilychen/miniforge3/envs/sklearn-dev/lib/python3.12/site-packages/array_api_compat/torch/__init__.py'>
y          = tensor([0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000....0000, 2.0000, 0.0000, 2.0000, 0.0000, 2.0000, 0.0000, 2.0000,
        0.0000, 2.0000, 0.0000, 2.0000, 0.0000, 7.7629])
y_np       = array([0.e+00, 0.e+00, 0.e+00, 0.e+00, 0.e+00, 0.e+00, 0.e+00, 0.e+00,
       0.e+00, 0.e+00, 0.e+00, 0.e+00, 0.e+00, ...2.e+00, 2.e+00, 2.e+00, 2.e+00, 2.e+00, 2.e+00, 2.e+00, 2.e+00,
       2.e+00, 2.e+00, 2.e+00, 2.e+00, 2.e+00, 1.e+05])

sklearn/utils/tests/test_stats.py:29: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/utils/stats.py:41: in _weighted_percentile
    sample_weight = xp.asarray(
        array      = tensor([0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
        0.0000, 0.0000, 0.0000, 0.0000....0000, 2.0000, 0.0000, 2.0000, 0.0000, 2.0000, 0.0000, 2.0000,
        0.0000, 2.0000, 0.0000, 2.0000, 0.0000, 7.7629])
        device     = None
        is_array_api_compliant = False
        percentile = 50
        sample_weight = tensor([0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000,
        1.8750, 0.0000, 1.8750, 0.0000....0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750,
        0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 0.0000])
        xp         = <sklearn.utils._array_api._NumPyAPIWrapper object at 0x10fd57b30>
sklearn/utils/_array_api.py:435: in asarray
    return numpy.asarray(x, dtype=dtype)
        copy       = None
        device     = None
        dtype      = dtype('float64')
        self       = <sklearn.utils._array_api._NumPyAPIWrapper object at 0x10fd57b30>
        x          = tensor([0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000,
        1.8750, 0.0000, 1.8750, 0.0000....0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750,
        0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 0.0000])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = tensor([0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000,
        1.8750, 0.0000, 1.8750, 0.0000....0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750,
        0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 0.0000])
dtype = dtype('float64')

    def __array__(self, dtype=None):
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        if dtype is None:
            return self.numpy()
        else:
>           return self.numpy().astype(dtype, copy=False)
E           RuntimeError: Numpy is not available

dtype      = dtype('float64')
self       = tensor([0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000,
        1.8750, 0.0000, 1.8750, 0.0000....0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 1.8750,
        0.0000, 1.8750, 0.0000, 1.8750, 0.0000, 0.0000])

../../../miniforge3/envs/sklearn-dev/lib/python3.12/site-packages/torch/_tensor.py:1089: RuntimeError
```

Reply:
This test is feeding a xp torch array to scikit-learn without enabling array API dispatch. array API dispatch is not expected to work when not enabled.

Reply:
But it's true that the error message you obtain is very confusing.

Reply:
Also note that `_weighted_percentile` is a private method. As such it does no input validation. Input validation should always be performed by public API calling  `_weighted_percentile` internally (e.g. the `_validate_data` method of an estimator, or the `check_array` calls in a public metric function).

Those input validation will transform the input arrays differently depending on wither array API dispatch is enabled or not. When enabled, the xp arrays will passthrough unchanged, but when not enabled, `np.asarray` will be used to attempt to convert any input first and getting a numpy array as a result. When not possible (e.g. calling `np.asarray()` on some Pytorch tensor allocated on a non-cpu device), that will fail (hopefully with an explicit error message) before ever calling into `_weighted_percentile`.